### PR TITLE
Fix GetString memory leak

### LINKED ISSUE #69: GetString leaks memory
PR incoming

### PR DESCRIPTION
**Pull request by markaren**
GetString eats memory. This PR fixes it.

Closes #69

### CODE REVIEW

**Comment by markaren** [comment]
Mhh. Indeed this fails with PyFMI, but works with FMI4j... Don't know what to do, because it does leak. `PyUnicode_AsEncodedString` returns a new reference, which is never decrefed.

**Comment by markaren** [comment]
I'm starting to think this could be a FMI4j bug? I need to check the standard on string handling. 

For refernce, this work with PyFMI
```cpp
PyObject* value = PyList_GetItem(refs, i);
            PyObject* strValue = PyUnicode_AsEncodedString(value, "utf-8", nullptr);
            char* src = PyBytes_AsString(strValue);
            char* dst = reinterpret_cast<char*>(malloc(strlen(src) + 1));
            strcpy_s(dst, strlen(src) + 1, src);
            values[i] = dst;
            Py_DECREF(strValue);
```

But I dunno if thats correct. It uses less memory in FMI4j, but still seams to leak.

**Comment by fcollonval** [comment]
Hey I don't think this a bug of FMI4J.

I have seen there is a built-in package for tracing Python allocation. Maybe you can get more inside with it?

https://docs.python.org/3/library/tracemalloc.html

Not: they is another one to correct too 😉 

https://github.com/NTNU-IHB/PythonFMU/blob/6c9daafc3fe0163e3031705bcf62cc15c6b3b041/pythonfmu/pythonfmu-export/headers/pythonfmu/PyException.hpp#L27

**Comment by fcollonval** [comment]
From the specification:

> The strings returned by fmi2GetString must be copied in the target environment because the allocated memory for these strings might be deallocated by the next call to any of the fmi2 interface functions or it might be an internal string buffer that is reused.

So from your latest proposal, could it be that the copied string `dst` are not released neither?

**Comment by markaren** [comment]
`dst` leaks for sure. 

So how I read your excerpt is that we must deallocate the string in _another_ FMI call. Say the next invocation of `GetString`. Ok. Makes the code more complex, but doable.

**Comment by markaren** [comment]
This should do it !

**Comment by markaren** [comment]
> oss << PyBytes_AsString(PyUnicode_AsEncodedString(pRepr, "utf-8", nullptr)); 

I'll fix that once #71 is merged

**Comment by markaren** [comment]
We should probably release a patch including the two latest bugfixes soon.

File: pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp [hunk 1]
```diff
@@ -418,7 +418,9 @@ void PySlaveInstance::handle_py_exception(const std::string& what, PyGILState_ST
         oss << what << "\n";
         if (pExcValue != nullptr) {
             PyObject* pRepr = PyObject_Repr(pExcValue);
-            oss << PyBytes_AsString(PyUnicode_AsEncodedString(pRepr, "utf-8", nullptr));
+            PyObject* pyStr = PyUnicode_AsEncodedString(pRepr, "utf-8", nullptr);
+            oss << PyBytes_AsString(pyStr);
+            Py_DECREF(pyStr);
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp]: I wonder if `oss.str()` work after this?

File: pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp [hunk 2]
```diff
@@ -436,6 +441,7 @@ void PySlaveInstance::handle_py_exception(const std::string& what, PyGILState_ST
 
 PySlaveInstance::~PySlaveInstance()
 {
+    clearStrBuffer();
```
Comment (fcollonval) [inline pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp]: We should move that call in `cleanPyObject` as it should be called when an error occurs too.

File: pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp (hunk 1 above)
Comment (fcollonval) [reply pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp]: The new test (https://github.com/NTNU-IHB/PythonFMU/blob/60a3ffce139ceb024e773bf840b2224faa912f4b/pythonfmu/tests/test_integration.py#L302) passed. So this seems ok.

Running the test locally with `pytest pythonfmu -sv` to see the output. I got:

```
../pythonfmu/tests/test_integration.py::test_integration_throw_py_error [FATAL] Fatal py exception encountered: [doStep] PyObject_CallMethod
RuntimeError()
PASSED
```

File: pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp [hunk 3]
```diff
@@ -302,6 +302,8 @@ void PySlaveInstance::GetBoolean(const cppfmu::FMIValueReference* vr, std::size_
 
 void PySlaveInstance::GetString(const cppfmu::FMIValueReference* vr, std::size_t nvr, cppfmu::FMIString* values) const
 {
+    clearStrBuffer();
```
Comment (fcollonval) [inline pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp]: I think this should be inside the lambda function (with the GIL captured) as the Python function `Py_DECREF` is called when clearing the buffer.

Comment (markaren) [reply pythonfmu/pythonfmu-export/cpp/PySlaveInstance.cpp]: Indeed. I was think while I did  that that it did not call any Python code..

**Pull request merged**